worldengine/draw.py

```diff
@@ -209,26 +393,93 @@ def draw_riversmap(world, target):
-    max_elev_land = mask.max()
-    elev_delta_land = max_elev_land - min_elev_land
+    for y in range(world.height):
+        for x in range(world.width):
+            target.set_pixel(x, y, (c[y, x], c[y, x], c[y, x], 255))
 
-    mask = numpy.ma.array(e, mask=numpy.logical_not(world.ocean))  # only ocean
-    min_elev_sea = mask.min()
-    max_elev_sea = mask.max()
-    elev_delta_sea = max_elev_sea - min_elev_sea
 
-    c = numpy.empty(e.shape, dtype=numpy.float)
-    c[numpy.invert(world.ocean)] = (e[numpy.invert(world.ocean)] - min_elev_land) * 127 / elev_delta_land + 128
-    c[world.ocean] = (e[world.ocean] - min_elev_sea) * 127 / elev_delta_sea
-    c = numpy.rint(c).astype(dtype=numpy.int32)  # proper rounding
+def draw_satellite(world, target):
+    ''' This draws a "satellit map" - a view of the generated planet as it may look from space '''
 
+    # Get an elevation mask where heights are normalized between 0 and 255
+    elevation_mask = get_normalized_elevation_mask(world)
+
+    ## The first loop sets each pixel's color based on colors defined in _biome_satellite_colors
+    #  and additional "business logic" defined in get_biome_color_based_on_elevation
     for y in range(world.height):
         for x in range(world.width):
-            target.set_pixel(x, y, (c[y, x], c[y, x], c[y, x], 255))
+            # Get the normalized elevation at this pixel
+            elev = elevation_mask[y, x]
+            
+            # Get a rgb noise value, with some logic to modify it based on the elevation of the tile
+            r, g, b = get_biome_color_based_on_elevation(world, elev, x, y)
+
+            # Set pixel to this color. This initial color will be accessed and modified later when 
+            # the map is smoothed and shaded.
+            target.set_pixel(x, y, (r, g, b, 255))
+
+
+    # Loop through and average a pixel with its neighbors to smooth transitions between biomes
+    for y in range(1, world.height-1):
+        for x in range(1, world.width-1):
+            ## Only smooth land tiles
+            if world.is_land((x, y)):
+                # Lists to hold the separated rgb values of the neighboring pixels
+                all_r = []
+                all_g = []
+                all_b = []
+
+                # Loop through this pixel and all neighboring pixels
+                for j in range(y-1, y+2):
+                    for i in range(x-1, x+2):
+                        # Don't include ocean in the smoothing, if this tile happens to border an ocean
+                        if world.is_land((i, j)):
+                            # Grab each rgb value and append to the list
+                            r, g, b, a = target.pixels[i, j]
+                            all_r.append(r)
+                            all_g.append(g)
+                            all_b.append(b)
+
+                # Making sure there is at least one valid tile to be smoothed before we attempt to average the values
+                if all_r:
```

Comment:
Would you mind doing the full check like 'len(...) > 0'?
I have seen some unexpected behavior due to this kind of abbreviated check.